core/src/consensus/stake/actions.rs

```diff
@@ -102,12 +116,68 @@ impl Action {
                     })?;
                 }
             }
+            Action::ReportDoubleVote {
+                message1,
+                message2,
+            } => {
+                if message1 == message2 {
+                    return Err(SyntaxError::InvalidCustomAction(String::from("messages are duplicated")))
+                }
+                if message1.round() != message2.round() {
+                    return Err(SyntaxError::InvalidCustomAction(String::from(
+                        "The messages are from two different voting rounds",
+                    )))
+                }
+
+                let signer_idx1 = message1.signer_index();
+                let signer_idx2 = message2.signer_index();
+
+                let signed_block_hash = message1.block_hash();
+                let client = client.expect("Client should be initialized");
+                let parent_block_hash = signed_block_hash
+                    .and_then(|hash| client.block_header(&hash.into()))
```

Comment:
Can we sure that we can get a block header from `message1.block_hash()`? We don't know whether `message1` is a vote to a valid block or an invalid. When `message1` is a vote to an invalid block, We couldn't get the header. We should impose a limitation that `message1` should be a valid vote and `message2` is an invalid vote, or make them symmetric, or get header from height.

Reply:
I think this also missing the case
```
message1 != message2 
&& message1.round() == message2.round()
&& message1.block_hash() == message2.block_hash()
&& signer(message1) != signer(message2)
```
Same block, Same round, but different signer (obviously they are different message)

Reply:
Thank you. I changed the code to get parent_block_hash based on the height information.